libcst/codemod/visitors/_gather_unused_imports.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+#
+
+from typing import Dict, List, Optional, Sequence, Set, Tuple, Union, Iterable
+
+import libcst as cst
+import libcst.matchers as m
+from libcst.codemod._context import CodemodContext
+from libcst.codemod._visitor import ContextAwareVisitor
+from libcst.metadata import QualifiedNameProvider, ScopeProvider, PositionProvider
+from libcst.codemod.visitors._gather_exports import GatherExportsVisitor
+from libcst.codemod.visitors._gather_comments import GatherCommentsVisitor
+from libcst.codemod.visitors._gather_string_annotation_names import (
+    GatherNamesFromStringAnnotationsVisitor,
+)
+from libcst.metadata.scope_provider import _gen_dotted_names
+
+
+DEFAULT_SUPPRESS_COMMENT_REGEX = (
+    r".*\W(lint-ignore: ?unused-import|noqa|lint-ignore: ?F401)(\W.*)?$"
+)
+
+
+class GatherUnusedImportsVisitor(ContextAwareVisitor):
+
+    SUPPRESS_COMMENT_REGEX_CONTEXT_KEY = f"GatherUnusedImportsVisitor.suppress_regex"
+    METADATA_DEPENDENCIES = (
+        *GatherCommentsVisitor.METADATA_DEPENDENCIES,
+        *GatherNamesFromStringAnnotationsVisitor.METADATA_DEPENDENCIES,
+        PositionProvider,
+        ScopeProvider,
+    )
+
+    def __init__(self, context: CodemodContext) -> None:
+        super().__init__(context)
+
+        self._string_annotation_names: Set[str] = set()
+        self._ignored_lines: Set[int] = set()
+        self._exported_names: Set[str] = set()
+        self.unused_imports: Set[
+            Tuple[cst.ImportAlias, Union[cst.Import, cst.ImportFrom]]
+        ] = set()
+
+    def visit_Module(self, node: cst.Module) -> bool:
+        export_collector = GatherExportsVisitor(self.context)
+        node.visit(export_collector)
+        self._exported_names = export_collector.explicit_exported_objects
+        comment_visitor = GatherCommentsVisitor(
+            self.context,
+            self.context.scratch.get(
+                self.SUPPRESS_COMMENT_REGEX_CONTEXT_KEY, DEFAULT_SUPPRESS_COMMENT_REGEX,
+            ),
```

Comment:
The thing with suppressing comments at `GatherUnusedImportsVisitor` level is that in some use cases (linter rules in frameworks that already silent lines with such comments) this ends up in doing more work than needed.

Maybe the suppression of comments could be moved to some other visitor which adds lines to be excluded, so one could choose whether to have them collected or not by using that one.

Reply:
I suppose you can set the regex to something that will never match (like `^$`), but we could add an explicit bool option to disable it. Would that help?

Reply:
Well, faster code is code never executed. Best thing would be if we could put these in a whole different, also reusable, place so they're never executed when they're not needed. Something like `GatherIgnoredLines` or something like that. But that's only a suggestion for those use cases where this isn't really needed.